### ISSUE TITLE
User should be able to deploy custom code for Infinispan

### ISSUE THREAD

**Issue by rigazilla** [infinispan/infinispan-operator#609]
The operator should support the deployment of custom code with the image.

Some possible approaches:
1. pack the custom code into an image and mount it as a sidecar that shares file system.
1. s2i image workflow could be applied on the solution above
1. provide a full s2i implementation that builds custom code on top of the standard Infinispan image.

The final product should work with OLM/Catalog/Hub.

ADDED:
Operator should take care of all the integration details, i.e.: deploying sidecar, mounting fs, configuring Infinispan

**Comment by ryanemerson** [comment]
My vote is for 1. The side-car approach provides a clean separation between the user dependencies and the base server, with k8 mounts linking the two. This also would have zero impact on the OLM manifest as we always use our images for operator deployments and the custom image(s) for the side-car. Users can provide any custom images via the Infinispan CR spec, with updates to the field triggerring a redeployment of the cluster.

Automating this process via 2 is also worth investigating, but we need to ensure we have a solution for both openshift and k8.

**Comment by dmvolod** [comment]
> * pack the custom code into an image and mount it as a sidecar that shares file system.

I'm not sure that this approach should be an easy than build the own image with docker/podman, push to the registry and use custom image option in the Infinispan CR. That should be measured between approaches for customer usability and spending time 
 
> * s2i image workflow could be applied on the solution above

This should be an optimal option
> * provide a full s2i implementation that builds custom code on top of the standard Infinispan image.

This can be too much complex, need to understand how many cases customer wants to use some additional code in image and possible code delivery options in real life (binary, source local or in git-based repository)

**Comment by rigazilla** [comment]
How about just mount a PV with the jars in?

**Comment by dmvolod** [comment]
> How about just mount a PV with the jars in?

it seems that you will need to create a Pod before putting the data to the PV
https://dzone.com/articles/kubernetes-copying-a-dataset-to-a-statefulsets-persistentvolume

**Comment by Crumby** [comment]
>> How about just mount a PV with the jars in?

> it seems that you will need to create a Pod before putting the data to the PV.

Providing extra single shared PV between all the instances by default could be an another option. User would just need to rsync the jars as it's not possible to put the data in PV before it's mount to a pod. Issue in here is that the server doesn't support hot deployments so we would need to trigger the deployment somehow. Maybe upon change in `spec.deployments.xyz`?

Creating `Deployment CR` would be probably taking heavy machinery for a simple task?

One more point to bare in mind is disconnected support. If we implement git source then deployment with git source won't be disconnected friendly.

**Comment by dmvolod** [comment]
> One more point to bare in mind is disconnected support. If we implement git source then deployment with git source won't be disconnected friendly.

Yes, sure. There are a lot of our customers who are using private git repositories installation inside and outside OCP, i.e. gitlab, gogs, MS Teams, etc.

**Comment by rigazilla** [comment]
> User would just need to rsync the jars as it's not possible to put the data in PV before it's mount to a pod. 

PV doesn't need to be empty, it could be prepared with the jars, pod just needs to mount it with all the jars already there.

**Comment by Crumby** [comment]
> pod just needs to mount it with all the jars already there

How do you get the jars into PV? As far as I'm aware of it's not possible to fill the PV withtout the PV being mounted to some pod at least for the time of CCying the data onto the PV. (I think you'd need at least a `job` that would populate the PV)

**Comment by rigazilla** [comment]
It depends on the storage provided, i.e. with NFS you can just copy your files in the exported folder. 
yeah, of course this means you must have some level of access to the storage infrastructure

**Comment by ryanemerson** [comment]
A PV based solution has two problems imo:

1. Adding the content to the volume in the first place. It's necessary to create, mount and then copy the file content over.
2. Step 1 is not portable. So it's necesary for step 1. to be repeated per cluster (namespace) or when a cluster is redeployed elsewhere.

An image based solution requires the image to be created only once (unless content requires updating) and is portable across environments. For offline deployments, it's just necessary for the image to be present in the local image repository.

**Comment by dmvolod** [comment]
> An image based solution requires the image to be created only once (unless content requires updating) and is portable across environments. For offline deployments, it's just necessary for the image to be present in the local image repository.

In this case don't see difference between custom image provided as right now via `Spec.Image` The same procedure, take Infinispan Server base image and add  a custom layer for the code. That's all, no more additional steps needed.

**Comment by Crumby** [comment]
> In this case don't see difference between custom image provided as right now via Spec.Image

`Spec.Image` doesn't work with updates and enables way more modifications then just adding additional jars.

**Comment by ryanemerson** [comment]
The problem with `Spec.Image` is that the user has to be concerned with the Infinispan server distribution. This means that when the user creates their custom image, they have to ensure that:

1. The base Infinispan server image they're using is compatible with the current operator version
2. They don't accidentally overwrite/delete any of the server content

Whereas utilising a separate image for deployments, allows user to use a scratch image which only contains their dependencies and prevents the user from accidentally modifying the Infinsipan server in a way that could break operator integration.

**Comment by dmvolod** [comment]
> > In this case don't see difference between custom image provided as right now via Spec.Image
> 
> `Spec.Image` doesn't work with updates and enables way more modifications then just adding additional jars.

Yes, sure, know about it. But it would be nice to have more intelligent way rather than package custom code/additional resources with docker/podman. It easy for you and others devs, but could be complex and not clear for customers

**Comment by ryanemerson** [comment]
> Yes, sure, know about it. But it would be nice to have more intelligent way rather than package custom code/additional resources with docker/podman. It easy for you and others devs, but could be complex and not clear for customers

I agree that a higher level abstraction would be nice, however I'm a little confused as your suggestion of using `Spec.Image` also requires these steps.

**Comment by dmvolod** [comment]
> I agree that a higher level abstraction would be nice, however I'm a little confused as your suggestion of using `Spec.Image` also requires these steps.

Sorry, it was not solution, it was an abstraction as this is slightly similar to provide custom image. But using custom code must be not the same as pack it inside image. It must be more intelligent way without using tools like docker, podman, etc. For example, if customer will use CodeReady Containers in the cloud, there is no way to use docker tools inside it or it can be not so easy way.

**Comment by rigazilla** [comment]
Looks like the only way to share things is passing through the registry.
Or do we have other options?
maybe git(hub) for s2i
maven?
NFS?

**Comment by dmvolod** [comment]
> maybe git(hub) for s2i

+1
git is common and standard for internet (github) and all disconnected things

**Comment by ryanemerson** [comment]
Resolving dependencies via maven is possible. I think it can also work offline if the [Nexus repository](https://blog.sonatype.com/nexus-repository-now-certified-in-red-hat-openshift-1) is installed.

The [camel-k operator](https://camel.apache.org/camel-k/latest/architecture/architecture.html) reconciles allows maven dependencies  and repositories to be configured via the [IntegrationSpec](https://camel.apache.org/camel-k/latest/architecture/cr/integration.html), these are then used by a builder image to retrieve/compile dependencies as required and create an executable image for deployment.

We could do something similar. Where we allow users to specify dependencies via the spec, the operator must then reconcile these dependencies before creating the cluster's StatefulSet. The compiled dependencies can then be mounted to a PersistentVolume, whilst provided images can be used consumed via sidecar. During an upgrade/restart, it's not necessasry to recreate the dependencies unless the spec has also changed.

Obviously we don't want to re-invent the wheel here, so we should leverage existing builder images/operators where possible.

As an initial idea, the `InfinispanSpec` could be updated to something like this:

```go
type InfinispanSpec struct {
	Replicas  int32                   `json:"replicas"`
	Image     *string                 `json:"image,optional,omitempty"`
        ...
        Dependencies InfinispanExternalDependencies  `json:"dependencies,optional,omitempty"`
}

type InfinispanExternalDependencies struct {
     Images []InfinispanImageDependency  `json:"images,optional,omitempty"`
     Jars []InfinispanMavenDependency  `json:"mavenDependencies,optional,omitempty"`
     Repositories []InfinispanGitDependency `json:"respositories,optional,omitempty"`
}

type InfinispanImageDependency struct {
    Image string
    SubPath string
}

type InfinispanMavenDependency {
    // Usual maven string groupId:artifact:version
    Url string
}

type InfinispanGitDependency {
   Url string
}
```

**Comment by ryanemerson** [comment]
For an initial implementation, I think `InfinispanImageDependency` and `InfinispanMavenDependency` would be the simplest to implement.

**Comment by ryanemerson** [comment]
I spent this morning looking through the `camel-k` code and there is a lot we can take inspiration from. Mainly how they handle the retrieval of dependencies and the creation of maven projects for single .java files. The [builder](https://github.com/apache/camel-k/tree/master/pkg/builder) is a good place to start.

Assuming we spin up a "builder" pod to initialize deployments, we could do the following:

# Maven Jar Dependencies
The deployment of existing jars, e.g. db drivers.

## Kubernetes
Utilise maven's built in CLI capabilities to download the specified maven artifacts from the configured repositories and mount them in the `server/lib` dir.

## Cli
```
// Create new cluster
kubectl infinispan create cluster [name] --dependency="mvn:org.postgresql:postgresql:42.2.17"

// Add dependency to existing cluster, trigerring a redeployment
kubectl infinispan cluster [name] --dependency="mvn:org.postgresql:postgresql:42.2.17"
```

# Source Dependencies
For source deployments, I think we can allow users to provide sources in two ways:
1. Source files mounted via a ConfigMap
2. A remote Git (GitHub) url

## Kubernetes
Retrieve the source via 1 or 2, compile with maven and copy the configured target jar(s) to a volume that is mounted in `server/lib`.

## Cli
```
// Local projects
--dependency="file:/some/local/dir"

// Git repo
--dependency="git@github.com:infinispan/infinispan.git"
```

camel-k utilise the `adoptopenjdk/openjdk11:slim` image for compilation and download maven via curl, however we can simply utilise the `ubi8/openjdk-11` image which also contains s2i integration scripts.

N.B. The camel-k code structure is very nice, with controller logic split across multiple files.

**Comment by dmvolod** [comment]
We can also provide very simple option for custom code deployment implemented in keycloak Operator
https://github.com/keycloak/keycloak-operator/blob/1f5896c5ddcdaa45da928e4fd935122e84069df1/pkg/apis/keycloak/v1alpha1/keycloak_types.go#L26-L29

**Comment by ryanemerson** [comment]
Closing. Users are now able to download code via http/ftp or deploy custom code via a PVC.
